### ISSUE TITLE
C++ API, "tried creating tensor with negative value in shape" error when 'permute' and 'reshape' functions are used

### ISSUE THREAD

**Issue by okankop** [microsoft/onnxruntime#11069]
I am using the C++ API of the onnxruntime lib, and I encountered "tried creating tensor with negative value in shape" error at the inference of the model. The model that I am using is a CRN architecture similar to https://web.cse.ohio-state.edu/~tan.650/doc/papers/Tan-Wang1.interspeech18.pdf.

I got down on the model architecture and figured out that the error arises from "permute" and "reshape" functions that are used to format the output of the encoder block that is used in the RNN layer. If I remove these functions (obviously the model inference is not performed as expected with this setting), I do not get any error

Note: I do not observe any errors if I use Python API.

**System information**
- Windows 11
- ONNX Runtime installed from source:
- ONNX Runtime version: 1.10.0

**Comment by yuslepukhin** [comment]
Are you feeding a shape with negative values when creating one or more of the inputs?

**Comment by okankop** [comment]
No, I always feed fixed size of input. Also, for the same input, if I remove reshape and permute function frim my model, I do not get any error.

**Comment by okankop** [comment]
[files_to_reproduce.zip](https://github.com/microsoft/onnxruntime/files/8397739/files_to_reproduce.zip)

Attached, I am providing the files to reproduce the error. If the model in 'model+dummy_working.py' is used, everything works perfectly. However, if we try to use model in 'model_dummy_error.py', we get an error in C++ API **"tried creating tensor with negative value in shape"**.

Both models run without problem in onnxruntime Python API.

**Comment by yuslepukhin** [comment]
The error in question is issued from the API entry point when the client code attempts to create inputs/outputs. So it is more likely to be related to the input data(shape) you are feeding and the related client code rather than the specific model itself.

**Comment by hongyangdengms** [comment]
The error happens when calling

OnnxModelRuntime.BindInputOutput( { InputBuffer }, { 1600 },
                                                              { OutputBuffer }, { 1600 }));

The inputBuffer and outputBuffer are one dimensional arrays with size 1600.

**Comment by okankop** [comment]
I was using the output dimensions returned from the onnx model. Apparently, for some models the returned output dimensions are -1. So hardcoding the expected output dimensions resolves the issue. So I am closing the issue here.

**Comment by yuslepukhin** [comment]
@okankop I would not recommend hardcoding the dimensions for the same reason you do not want to pass expected dimensions from metadata. By doing so you assume that your shape always matches your allocation and that disables a helpful check that Onnxruntime can perform for you. The shape that you pass should be computed together with doing the actual allocation and be passed to the API.

**Comment by okankop** [comment]
@yuslepukhin this was the reason why I was trying to get the output shape from the model and create the output tensor accordingly. However, GetShape() function returns -1 when I use 'permute' and 'reshape' functions in the model. Here is my failing code snippet:
`    // Prepare outputs`
`    for (size_t idx = 0; idx < NumOutputNodes; idx++) {`
    ` const std::vector<int64_t> OutputNodeDims = m_Session->GetOutputTypeInfo(idx).GetTensorTypeAndShapeInfo().GetShape();`
`        m_OutputNodeNames.push_back(m_Session->GetOutputName(idx, m_Allocator));`
`        m_OutputTensors.push_back(Ort::Value::CreateTensor<float>(MemoryInfo, OutputBuffer[idx], OutputSize[idx],`
`                                                                  OutputNodeDims.data(),`
`                                                                  OutputNodeDims.size()));`
`    }`

Do you have any suggestions to modify the above code snippet to avoid this error?
